### PR TITLE
[Api] Locale code on checkout

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 1]
```diff
@@ -716,11 +722,14 @@ public function thereShouldBeNoTaxesCharged(): void
...
+     * @Then my order's locale should be :locale
      */
-    public function myOrderLocaleShouldBe(string $localeCode): void
+    public function myOrderLocaleShouldBe(LocaleInterface $locale): void
     {
-        Assert::same($this->responseChecker->getValue($this->ordersClient->getLastResponse(), 'localeCode'), $localeCode);
+        Assert::same($this
+            ->responseChecker
+            ->getValue($this->ordersClient->getLastResponse(), 'localeCode'), $locale->getCode()
+        );
```
Comment (arti0090) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:732]:
```suggestion
        Assert::same($this->responseChecker->getValue(
            $this->ordersClient->getLastResponse(), 'localeCode'),
            $locale->getCode()
        );
```

I feel that this is more elegant and better to read, but I might be wrong 😄

**Review by arti0090 — APPROVED** [review]

**Review by AdamKasp — APPROVED** [review]

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php (hunk 1 above)
Comment (Tomanhez) [reply src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:732]: We should 
```suggestion
        Assert::same(
            $this->responseChecker->getValue(
                $this->ordersClient->getLastResponse(), 'localeCode'
            ),
            $locale->getCode()
        );
```

Possibly we should change it that, your proposition is ok but in first line we have opened two parenthesis and should be closed in the same next lines. In your case second parenthesis is closed in second line and first parenthesis is closed in fourth line. We should remember about this rules 🚀

Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:732]: Why did you change it?

File: src/Sylius/Component/Core/Model/Channel.php [hunk 2]
```diff
@@ -180,6 +180,17 @@ public function hasLocale(LocaleInterface $locale): bool
...
+    public function hasLocaleWithLocaleCode(string $localeCode): bool
+    {
+        return !$this
+            ->locales
+            ->filter(function (LocaleInterface $locale) use ($localeCode): bool {
+                return $locale->getCode() === $localeCode;
+            })
+            ->isEmpty()
+        ;
+    }
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Model/Channel.php:192]: It is used only once. Can't we implement out of the object?

File: src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php [hunk 3]
```diff
@@ -85,11 +85,22 @@ public function __invoke(PickupCart $pickupCart)
...
+
+        /** @var string $localeCode */
+        $localeCode = $locale->getCode();
+
+        /** @var string|null $commandLocaleCode */
+        $commandLocaleCode = $pickupCart->localeCode;
+
+        if ($commandLocaleCode !== null && $channel->hasLocaleWithLocaleCode($commandLocaleCode)) {
+            $localeCode = $commandLocaleCode;
+        }
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php:97]:
```suggestion
        /** @var string|null $commandLocaleCode */
        $localeCode = $pickupCart->localeCode;

        if ($$localeCode !== null && !$channel->hasLocaleWithLocaleCode($localeCode)) {
            /** @var LocaleInterface $locale */
            $locale = $channel->getDefaultLocale();
            
            $localeCode = $locale->getCode();
        }
```

This should be also documented on API endpoint

File: features/checkout/seeing_order_summary/seeing_order_locale.feature [hunk 4]
```diff
@@ -19,9 +19,17 @@ Feature: Seeing order locale on order summary page
...
         When I proceed through checkout process in the "French (France)" locale
         Then I should be on the checkout summary step
-        And my order's locale should be "français (France)"
+        And my order's locale should be "French (France)"
+
+    @api
+    Scenario: Seeing order locale on the order summary page after choosing channel locale
+        Given I have cart picked up in the "French (France)" locale
+        And I have product "Stark T-Shirt" in the cart
+        When I proceed through checkout process
```
Comment (lchrusciel) [inline features/checkout/seeing_order_summary/seeing_order_locale.feature:33]: It should rather check if we can pickup-cart in locale through API. Also, in the long term, we should have a validation check, if the locale is available during update

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php (hunk 1 above)
Comment (Tomanhez) [reply src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:732]: In @ui tests was using locale name, but in @api we need localeCode, It worked before, because test for `en_US` working properly for both, but to use fr_FR created problems

File: src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php [hunk 5]
```diff
@@ -111,4 +119,16 @@ private function provideCustomer(): ?CustomerInterface
...
+    private function hasLocaleWithLocaleCode(ChannelInterface $channel, ?string $localeCode)
+    {
+        $locales = $channel->getLocales();
+
+        return !$locales
+            ->filter(function (LocaleInterface $locale) use ($localeCode): bool {
+                return $locale->getCode() === $localeCode;
+            })
+            ->isEmpty()
+        ;
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php:132]: My recommendation, to improve readibility:

```suggestion
        $localeWithExcpectedCode = $locales->filter(function (LocaleInterface $locale) use ($localeCode): bool {
             return $locale->getCode() === $localeCode;
        });
        
        return !$localeWithExcpectedCode->isEmpty();
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml [hunk 6]
```diff
@@ -39,7 +39,7 @@
                     <attribute name="groups">order:create</attribute>
                 </attribute>
                 <attribute name="openapi_context">
-                    <attribute name="summary">Pickups a new cart</attribute>
+                    <attribute name="summary">Pickups a new cart. You can choose localeCode</attribute>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:42]:
```suggestion
                    <attribute name="summary">Pickups a new cart. Provided locale code has to be one of available for a particular channel. Setting different local code will result in the assignment of the default locale from the channel.</attribute>
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml [hunk 7]
```diff
@@ -111,6 +111,8 @@
         <attribute name="localeCode">
             <group>shop:order:account:read</group>
             <group>checkout:read</group>
+            <group>cart:update</group>
+            <group>order:read</group>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml:115]: As you are touching this area, can we introduce `shop:` prefix to order read serialization? Can be done in separate PR

**Comment by lchrusciel** [comment]
Rebase needed

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 8]
```diff
@@ -144,10 +145,11 @@ public function iRemoveProductFromTheCart(ProductInterface $product, string $tok
 
     /**
      * @When I pick up my cart (again)
+     * @When I pick up cart in the :localeCode locale
      */
-    public function iPickUpMyCart(): void
+    public function iPickUpMyCart(string $localeCode = null): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:150]:
```suggestion
    public function iPickUpMyCart(?string $localeCode = null): void
```

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 9]
```diff
@@ -158,6 +160,17 @@ public function iCheckDetailsOfMyCart(string $tokenValue): void
...
 
+    /**
+     * @Then my cart's locale should be :locale
+     */
+    public function myCartLocaleShouldBe(LocaleInterface $locale): void
+    {
+        Assert::same($this
+            ->responseChecker
+            ->getValue($this->cartsClient->getLastResponse(), 'localeCode'), $locale->getCode()
+        );
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:171]:
```suggestion
        Assert::same($this->responseChecker->getValue(
            $this->cartsClient->getLastResponse(), 'localeCode'), 
            $locale->getCode()
        );
```

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 10]
```diff
@@ -426,9 +439,10 @@ public function iShouldHaveEmptyCart(string $tokenValue): void
         Assert::same(count($items), 0, 'There should be an empty cart');
     }
 
-    private function pickupCart(): string
+    private function pickupCart(string $localeCode = null): string
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:442]:
```suggestion
    private function pickupCart(?string $localeCode = null): string
```

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php (hunk 1 above)
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:732]: Could still be in one line, or similar to my suggestion from above comment

File: src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php [hunk 11]
```diff
@@ -83,13 +83,21 @@ public function __invoke(PickupCart $pickupCart)
...
-        $locale = $channel->getDefaultLocale();
+        /** @var string|null $localeCode */
+        $localeCode = $pickupCart->localeCode;
+
+        if ($localeCode === null || !$this->hasLocaleWithLocaleCode($channel, $localeCode)) {
+            /** @var LocaleInterface $locale */
+            $locale = $channel->getDefaultLocale();
+
+            $localeCode = $locale->getCode();
+        }
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php:94]: Probably this logic could be extracted to a separate private method

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml (hunk 7 above)
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml:116]: Ok I do it in new PR after merge this

**Review by arti0090 — APPROVED** [review]

File: features/checkout/seeing_cart_locale.feature [hunk 12]
```diff
@@ -0,0 +1,16 @@
+@checkout
+Feature: Seeing cart locale on cart details
```
Comment (lchrusciel) [inline features/checkout/seeing_cart_locale.feature:2]:
```suggestion
Feature: Picking up the cart with the locale other than the default
```

File: features/checkout/seeing_cart_locale.feature [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+    In order be certain what is the cart locale
```
Comment (lchrusciel) [inline features/checkout/seeing_cart_locale.feature:3]:
```suggestion
    In order to make shopping the most convenient way
```

File: features/checkout/seeing_cart_locale.feature [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+    As a Customer
+    I want to be able to see cart locale on the cart details
```
Comment (lchrusciel) [inline features/checkout/seeing_cart_locale.feature:5]:
```suggestion
    I want to be able to make shopping in the most preferable language from available
```

File: features/checkout/seeing_cart_locale.feature [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And that channel allows to shop using "English (United States)" and "French (France)" locales
+        And I am a logged in customer
+
+    @api
+    Scenario: Seeing cart locale on the cart summary page after choosing channel locale
```
Comment (lchrusciel) [inline features/checkout/seeing_cart_locale.feature:13]:
```suggestion
    Scenario: Picking up the cart with the locale other than default
```

**Review by lchrusciel — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml [hunk 16]
```diff
@@ -39,7 +39,7 @@
                     <attribute name="groups">order:create</attribute>
                 </attribute>
                 <attribute name="openapi_context">
-                    <attribute name="summary">Pickups a new cart</attribute>
+                    <attribute name="summary">Pickups a new cart. Provided locale code has to be one of available for a particular channel. Setting different local code will result in the assignment of the default locale from the channel.</attribute>
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:42]:
```suggestion
                    <attribute name="summary">Pickups a new cart. Provided locale code has to be one of available for a particular channel. Setting different locale code will result in the assignment of the default locale from the channel.</attribute>
```

Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:42]: is it proper behaviour? IMHO if we provided locale which isn't in the channel we should get an error, not default locale...

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml [hunk 17]
```diff
@@ -306,6 +306,17 @@
...
+
+            <itemOperation name="shop_put">
+                <attribute name="method">PUT</attribute>
+                <attribute name="path">/shop/orders/{id}</attribute>
+                <attribute name="denormalization_context">
+                    <attribute name="groups">cart:update</attribute>
+                </attribute>
+                <attribute name="openapi_context">
+                    <attribute name="summary">Change locale code on cart</attribute>
+                </attribute>
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:318]: IMHO it is redundant, we don't need to explain normal `put` endpoint

Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:318]: But now it's only available parameter on order put method which you can change.

Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:318]: still, it is normal `put`, on, for example, product `put` will you list all available parameter?

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml [hunk 18]
```diff
@@ -39,7 +39,7 @@
                     <attribute name="groups">order:create</attribute>
                 </attribute>
                 <attribute name="openapi_context">
-                    <attribute name="summary">Pickups a new cart</attribute>
+                    <attribute name="summary">Pickups a new cart. Provided locale code has to be one of available for a particular channel. Setting different locale code will result in the assignment of the default locale from the channel.</attribute>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:42]:
```suggestion
                    <attribute name="summary">Pickups a new cart. Provided locale code has to be one of available for a particular channel.</attribute>
```

**Review by lchrusciel — APPROVED** [review]

**Comment by lchrusciel** [comment]
Thanks, Tomasz! :tada:
